File: truss-compiler/src/main/java/truss/compiler/Span.java
package truss.compiler;

public class Span {
    public static final String UNKNOWN_FILENAME = "(unknown)";
    public static final Span EMPTY = new Span(null);

    private final String fileName;
    private final int startLine;
    private final int startColumn;
    private final int endLine;
    private final int endColumn;

    public Span(String fileName) {
        this.fileName = fileName;
        startLine = -1;
        startColumn = -1;
        endLine = -1;
        endColumn = -1;
    }

    public Span(String fileName, int startLine, int startColumn) {
        this(fileName, startLine,  startColumn, startLine, startColumn);
    }

    public Span(String fileName, int startLine, int startColumn, int endLine, int endColumn) {
        this.fileName = fileName;
        this.startLine = startLine;
        this.startColumn = startColumn;
        this.endLine = endLine;
        this.endColumn = endColumn;
    }

    public String getFileName() {
        return fileName;
    }

    public int getStartLine() {
        return startLine;
    }

    public int getStartColumn() {
        return startColumn;
    }

    public int getEndLine() {
        return endLine;
    }

    public int getEndColumn() {
        return endColumn;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        if (fileName == null) {
            sb.append(UNKNOWN_FILENAME);
        } else {
            sb.append(fileName);
        }

        if (startLine != -1 && startColumn != -1) {
            sb.append('(');
            sb.append(startLine);
            sb.append(',');
            sb.append(startColumn + 1);
            if (endLine != -1 && endColumn != -1 && (startLine != endLine || startColumn != endColumn)) {
                sb.append(',');
                sb.append(endLine);
                sb.append(',');
                sb.append(endColumn + 1);
            }
            sb.append(')');
        }

        return sb.toString();
    }
}

File: Truss.Compiler/Span.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Truss.Compiler {
    public class Span {
        public const string UnknownFilename = "(unknown)";
        public static readonly Span Empty = new Span(null);

        public Span(string fileName) {
            FileName = fileName;
            StartLine = -1;
            StartColumn = -1;
            EndLine = -1;
            EndColumn = -1;
        }

        public Span(string fileName, int startLine, int startColumn)
            : this(fileName, startLine, startColumn, startLine, startColumn) {
        }

        public Span(string fileName, int startLine, int startColumn, int endLine, int endColumn) {
            FileName = fileName;
            StartLine = startLine;
            StartColumn = startColumn;
            EndLine = endLine;
            EndColumn = endColumn;
        }

        public string FileName { get; private set; }

        public int StartLine { get; private set; }

        public int StartColumn { get; private set; }

        public int EndLine { get; private set; }

        public int EndColumn { get; private set; }

        public override string ToString() {
            var sb = new StringBuilder();

            sb.Append(FileName ?? UnknownFilename);

            if (StartLine != -1 && StartColumn != -1) {
                sb.Append('(');
                sb.Append(StartLine);
                sb.Append(',');
                sb.Append(StartColumn + 1);
                if (EndLine != -1 && EndColumn != -1 && (StartLine != EndLine || StartColumn != EndColumn)) {
                    sb.Append(',');
                    sb.Append(EndLine);
                    sb.Append(',');
                    sb.Append(EndColumn + 1);
                }
                sb.Append(')');
            }

            return sb.ToString();
        }
    }
}
